File: cli/src/commands/notifications.ts
import { ParsedArgs, flagString, flagBoolean } from "../args.js";
import { getClient, run } from "../client.js";
import { printTable, printJson, success, error, confirm } from "../output.js";

/**
 * "Notifications" maps to RelayHub's alert rules (/v1/alerts/*) -- Slack,
 * Discord, webhook, or email notifications fired on a threshold condition
 * (endpoint down, DLQ spike, high latency, etc). See sdks/node/README.md and
 * docs/api/notifications.md for the same note.
 */
export async function notificationsCommand(args: ParsedArgs): Promise<void> {
  await run(async () => {
    const client = getClient({ apiKey: flagString(args.flags, "api-key"), baseUrl: flagString(args.flags, "base-url") });
    const action = args.positionals[0] ?? "list";

    switch (action) {
      case "list": {
        const rules = await client.notifications.listRules();
        if (args.flags.json) return printJson(rules);
        printTable(
          ["ID", "CONDITION", "SEVERITY", "CHANNEL", "ENABLED"],
          rules.map((r) => [r.id, r.condition_type, r.severity, r.channel, r.is_enabled ? "yes" : "no"])
        );
        return;
      }
      case "create": {
        const conditionType = flagString(args.flags, "condition");
        const channel = flagString(args.flags, "channel");
        if (!conditionType || !channel) {
          return error("Usage: relay notifications create --condition <type> --channel slack|discord|webhook|email --channel-config '<json>' [--severity info|warning|critical]");
        }
        const channelConfigRaw = flagString(args.flags, "channel-config");
        const rule = await client.notifications.createRule({
          condition_type: conditionType,
          severity: (flagString(args.flags, "severity") as "info" | "warning" | "critical" | undefined) ?? "warning",
          channel,
          channel_config: channelConfigRaw ? JSON.parse(channelConfigRaw) : {},
        });
        success(`Created alert rule ${rule.id}`);
        return;
      }
      case "test": {
        const id = args.positionals[1];
        if (!id) return error("Usage: relay notifications test <ruleId>");
        const result = await client.notifications.testRule(id);
        if (result.delivery_status === "sent" || result.delivery_status === "success") {
          success(`Test notification sent (status: ${result.delivery_status})`);
        } else {
          error(`Test notification failed: ${result.delivery_status}${result.delivery_error ? " -- " + result.delivery_error : ""}`);
        }
        return;
      }
      case "delete": {
        const id = args.positionals[1];
        if (!id) return error("Usage: relay notifications delete <ruleId>");
        if (!flagBoolean(args.flags, "yes") && !(await confirm(`Delete alert rule ${id}?`))) return;
        await client.notifications.deleteRule(id);
        success(`Deleted alert rule ${id}`);
        return;
      }
      case "history": {
        const events = await client.notifications.history();
        if (args.flags.json) return printJson(events);
        printTable(
          ["ID", "CONDITION", "SEVERITY", "TRIGGERED", "DELIVERY STATUS"],
          events.map((e) => [e.id, e.condition_type, e.severity, e.triggered_at, e.delivery_status])
        );
        return;
      }
      default:
        error(`Unknown subcommand: notifications ${action}`);
        console.log("Usage: relay notifications <list|create|test|delete|history> [args]");
    }
  });
}

File: sdks/go/relayhub/notifications.go
package relayhub

import "context"

// NotificationsService maps to RelayHub's alert rules: Slack, Discord, webhook,
// or email notifications fired when an endpoint's failure rate crosses a
// threshold. There is no separate /notifications API in the backend -- this
// service is named for the developer-facing concept it serves, while every
// method below hits the real /v1/alerts/* routes.
type NotificationsService struct{ t *transport }

type CreateAlertRuleRequest struct {
	ConditionType         string         `json:"condition_type"`
	Severity              string         `json:"severity,omitempty"`
	Channel               string         `json:"channel"`
	ChannelConfig         map[string]any `json:"channel_config"`
	ThresholdConfig       map[string]any `json:"threshold_config,omitempty"`
	ThrottleWindowMinutes *int           `json:"throttle_window_minutes,omitempty"`
	IsEnabled             *bool          `json:"is_enabled,omitempty"`
}

// CreateRule calls POST /v1/alerts/rules.
func (s *NotificationsService) CreateRule(ctx context.Context, req CreateAlertRuleRequest, opts ...RequestOption) (AlertRule, error) {
	return decode[AlertRule](s.t.do(ctx, "POST", "/v1/alerts/rules", req, opts...))
}

// ListRules calls GET /v1/alerts/rules.
func (s *NotificationsService) ListRules(ctx context.Context, opts ...RequestOption) ([]AlertRule, error) {
	return decode[[]AlertRule](s.t.do(ctx, "GET", "/v1/alerts/rules", nil, opts...))
}

type UpdateAlertRuleRequest struct {
	Severity              string         `json:"severity,omitempty"`
	Channel               string         `json:"channel,omitempty"`
	ChannelConfig         map[string]any `json:"channel_config,omitempty"`
	ThresholdConfig       map[string]any `json:"threshold_config,omitempty"`
	ThrottleWindowMinutes *int           `json:"throttle_window_minutes,omitempty"`
	IsEnabled             *bool          `json:"is_enabled,omitempty"`
}

// UpdateRule calls PATCH /v1/alerts/rules/{id}.
func (s *NotificationsService) UpdateRule(ctx context.Context, ruleID string, req UpdateAlertRuleRequest, opts ...RequestOption) (AlertRule, error) {
	return decode[AlertRule](s.t.do(ctx, "PATCH", "/v1/alerts/rules/"+ruleID, req, opts...))
}

// DeleteRule calls DELETE /v1/alerts/rules/{id} (204 No Content on success).
func (s *NotificationsService) DeleteRule(ctx context.Context, ruleID string, opts ...RequestOption) error {
	_, err := s.t.do(ctx, "DELETE", "/v1/alerts/rules/"+ruleID, nil, opts...)
	return err
}

// TestRule calls POST /v1/alerts/rules/{id}/test -- fires a test notification through the rule's configured channel.
func (s *NotificationsService) TestRule(ctx context.Context, ruleID string, opts ...RequestOption) (TestAlertResponse, error) {
	return decode[TestAlertResponse](s.t.do(ctx, "POST", "/v1/alerts/rules/"+ruleID+"/test", nil, opts...))
}

// History calls GET /v1/alerts/history.
func (s *NotificationsService) History(ctx context.Context, conditionType string, opts ...RequestOption) ([]AlertEvent, error) {
	allOpts := opts
	if conditionType != "" {
		allOpts = append(allOpts, WithQuery("condition_type", conditionType))
	}
	return decode[[]AlertEvent](s.t.do(ctx, "GET", "/v1/alerts/history", nil, allOpts...))
}
